---
title: "Combining Deep Learning and TCRs with Trex"
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
output: rmarkdown::html_vignette
theme: united
df_print: kable
vignette: >
  %\VignetteIndexEntry{Combining Deep Learning and TCRs with Trex}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
***

<style>
p.caption {
  font-size: 0.9em;
}
</style>
 
```{r setup, include=FALSE}
all_times <- list()  # store the time for each chunk
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
      now <<- Sys.time()
    } else {
      res <- difftime(Sys.time(), now, units = "secs")
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  time_it = TRUE
)

suppressMessages(library(scRepertoire))
suppressMessages(library(Trex))
suppressMessages(library(Seurat))
suppressMessages(library(ggplot2))
suppressMessages(library(viridis))
suppressMessages(library(dplyr))
suppressMessages(library(reticulate))
use_condaenv(condaenv = "r-reticulate", required = TRUE)

data("contig_list") 
combined.TCR <- combineTCR(contig_list, 
                           samples = c("P17B", "P17L", "P18B", "P18L", 
                                       "P19B","P19L", "P20B", "P20L"), 
                           filterMulti = TRUE)

googledrive::drive_download("https://drive.google.com/file/d/1_YuRraDyg8UgF3oasjF0-jgPnwox-B24/view?usp=share_link", overwrite = TRUE)

scRep_example <- readRDS("scRep_example_full.rds")

scRep_example <- combineExpression(combined.TCR, 
                                   scRep_example, 
                                   cloneCall="gene", 
                                   group.by = "sample")

#Adding patient information
scRep_example$Patient <- substr(scRep_example$orig.ident, 1,3)

#Adding type information
scRep_example$Type <- substr(scRep_example$orig.ident, 4,4)
```


Moving beyond the dual-embedding of TCR/RNA using Trex, the approach can also be used for performing trajectory analysis. Unlike using RNA alone, we can leverage using the clonal identity as part of the calculation. For this vignette, we will there are several packages that will be used:   

* [mumosa(https://bioconductor.org/packages/release/bioc/html/mumosa.html) for neighbor rescaling.    
* [phateR](https://github.com/KrishnaswamyLab/phateR) for the diffusion-based affinity embedding.
* [slingshot](https://www.bioconductor.org/packages/release/bioc/html/slingshot.html) for lineage estimation and pseudotime calculation.  


## Loading Libraries

```{r}
library(mumosa)
library(phateR)
library(slingshot)
```

## Getting the Clonal Embeddings for TRA and TRB chains

[Trex](https://github.com/ncborcherding/Trex) takes the CDR3 amino acid sequence and embeds them into latent dimensional space using convolutional neural networks. An introduction to Trex and the models available can be found in the [scRepertoire website](https://www.borch.dev/uploads/screpertoire/articles/trex). Below we will use variational autoencoders (VAE) using the the Kidera factors.

For the purpose of this vignette, we will use Patent 17 samples and look specifically at CD4+ cells. Low expression of CD4 mRNA is commonly seen in single-cell RNA sequencing, so we will use the absence of *CD8A* and *CD8B* as a proxy for CD8 cells. 

```{r tidy=FALSE}
#Subsetting Patient Samples
SeuratObj <- subset(scRep_example, Patient == "P17")

#Subsetting CD4 Cells
"%!in%" <- Negate("%in%")
CD8A.pos <- rownames(SeuratObj[[]])[which(SeuratObj@assays$RNA@counts["CD8A",] > 1)]
CD8B.pos <- rownames(SeuratObj[[]])[which(SeuratObj@assays$RNA@counts["CD8B",] > 1)]
CD8.cells <- unique(CD8A.pos, CD8B.pos)
CD4.cells <- rownames(SeuratObj[[]])[rownames(SeuratObj[[]]) %!in% CD8.cells]
SeuratObj <- subset(SeuratObj, cells = CD4.cells)

######################
#Getting RNA Reductions
#######################

SeuratObj <- SeuratObj %>%
              NormalizeData(verbose = FALSE) %>%
              FindVariableFeatures(verbose = FALSE) %>%
              quietTCRgenes() %>%
              ScaleData(verbose = FALSE) %>%
              RunPCA(verbose = FALSE)

######################
#Getting TCR Reductions
#######################

SeuratObj <- runTrex(SeuratObj, 
                     chains = "TRA",
                     encoder.model = "VAE", 
                     encoder.input = "KF",
                     reduction.name = "Trex.TRA.KF")

SeuratObj <- runTrex(SeuratObj, 
                     chains = "TRB",
                     encoder.model = "VAE", 
                     encoder.input = "KF",
                     reduction.name = "Trex.TRB.KF")

plot1 <-  DimPlot(SeuratObj, 
                  reduction = "Trex.TRA.KF", 
                  group.by = "orig.ident") + 
              scale_color_viridis(option = "B", discrete = TRUE)+ 
              theme(plot.title = element_blank())

plot2 <-  DimPlot(SeuratObj, 
                  reduction = "Trex.TRB.KF", 
                  group.by = "orig.ident") + 
              scale_color_viridis(option = "B", discrete = TRUE) + 
              theme(plot.title = element_blank())

plot1 + plot2
```


## Creating RNA/TCR Clusters and UMAPs


```{r tidy=FALSE}
rescaled.values <- rescaleByNeighbors(
                            list(RNA = SeuratObj@reductions$pca@cell.embeddings,
                                 TRA = SeuratObj@reductions$Trex.TRA.KF@cell.embeddings[,1:3], 
                                 TRB = SeuratObj@reductions$Trex.TRB.KF@cell.embeddings[,1:3]))

y <- DiffusionMap(rescaled.values)
plot(y)

phate.coord <- phate(rescaled.values)

ggplot(phate.coord) +
  geom_point(aes(PHATE1, PHATE2, color = SeuratObj$CTaa)) + 
  guides(color = "none")


mds <- as.data.frame(calculateMDS(t(rescaled.values)))
ggplot(mds) +
  geom_point(aes(mds[,1], mds[,2], color = SeuratObj$CTaa)) + 
  guides(color = "none")

umap <- as.data.frame(calculateUMAP(t(rescaled.values)))
ggplot(umap) +
  geom_point(aes(umap[,1], umap[,2], color = SeuratObj$CTaa)) + 
  guides(color = "none")

umap.trimodal <- calculateMultiUMAP(
                            list(RNA = SeuratObj@reductions$pca@cell.embeddings,
                                 TRA = SeuratObj@reductions$Trex.TRA.KF@cell.embeddings[,1:3], 
                                 TRB = SeuratObj@reductions$Trex.TRB.KF@cell.embeddings[,1:3]), 
                           # metric = "cosine",
                            n_component = 25)
SeuratObj[["triUMAP"]] <- CreateDimReducObject(embeddings = umap.trimodal,
                                               key = "triUMAP")

SeuratObj <- SeuratObj %>%
              FindNeighbors(reduction = "triUMAP", 
                            dims = 1:20) %>%
              FindClusters(algorithm = 4, 
                           resolution = 0.1)

DimPlot(SeuratObj, reduction = "triUMAP")
DimPlot(SeuratObj, group.by = "CTaa", reduction = "triUMAP") + 
  theme(plot.title = element_blank()) + 
  scale_color_viridis(option = "B", discrete = TRUE) + 
  NoLegend()
```






